application/models/Os_model.php

```diff
@@ -309,4 +309,20 @@ public function criarTextoWhats($textoBase, $troca)
         $textoBase = htmlentities(urlencode($textoBase));
         return $textoBase;
     }
+
+    public function valorTotalOS($servicos, $produtos)
+    {
+        $totalServico = 0;
+        $totalProdutos = 0; 
+        foreach ($produtos as $p) {
+            $totalProdutos = $totalProdutos + $p->subTotal;
+        }
+        foreach ($servicos as $s) {
+            $preco = $s->preco ?: $s->precoVenda;
+            $subtotal = $preco * ($s->quantidade ?: 1);
+            $totalServico = $totalServico + $subtotal;
+        }
+
+        return ['totalServico' => $totalServico, 'totalProdutos' => $totalProdutos];
+    }
```

Comment:
Eu acho que essa função ao invés de receber os produtos e serviços como entradam o correto seria receber somente o id da OS e com o id você faz a consulta e o cálculo.

Se conseguir fazer será bom, vou deixar meu approve e fazer o merge mais tarde se você não quiser fazer.

Reply:
Está um pouco estranho também no for de produtos fazer com o subtotal, se puder fazer usando a quantidade vezes o preço mesmo creio que fique mais claro.

Pode usar também o array_reduce ao invés de usar for.

Reply:
Certo, eu aproveitei a entrada dos produtos e dos serviços por que já são consultas já realizadas, não teria que adicionar + uma, vou melhorar